src/core/config.c

```diff
@@ -0,0 +1,139 @@
+#include "config.h"
+#include "xdpw.h"
+
+#include <dictionary.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+#include <wordexp.h>
+#include "iniparser.h"
+
+#define SYSCONFDIR "/etc"
+
+static const char *loglevels[] = {
+	"UNSET",
+	"QUIET",
+	"ERROR",
+	"WARN",
+	"INFO",
+	"DEBUG",
+	"TRACE"
+};
+
+void print_config(enum LOGLEVEL loglevel, struct xdpw_config *config) {
+	logprint(loglevel, "config: Configfile %s",config->conf.configfile);
+	logprint(loglevel, "config: Loglevel %d",config->logger_conf.loglevel);
+	logprint(loglevel, "config: Outputname  %s",config->screencast_conf.output_name);
+}
+
+void free_configstring(char **dest) {
+	if (*dest != NULL) {
+		free(*dest);
+		*dest = NULL;
+	}
+}
+
+void destroy_config(struct xdpw_config *config) {
+	free_configstring(&config->conf.configfile);
+
+	// screencast
+	free_configstring(&config->screencast_conf.output_name);
+}
+
+void getstring_from_conffile(dictionary *d, char *key, char **dest, char *def) {
+	if (*dest == NULL) {
+		const char *c = iniparser_getstring(d, key, def);
+		if (c != NULL) {
+			// Allow keys without value as default
+			if (strcmp(c, "") != 0) {
+				*dest = strdup(c);
+			} else {
+				*dest = strdup(def);
+			}
+		}
+	}
+}
+
+void getenum_from_conffile(dictionary *d, char *key, int *dest, const char *array[] , int num_array, int def) {
+	if (*dest == 0) {
+		const char *c = iniparser_getstring(d, key, NULL);
+		if (c != NULL) {
+			for (int i = 0; i < num_array; i++) {
+				if (strcmp(c, array[i]) == 0) {
+					*dest = i;
+					return;
+				}
+			}
+		}
+		*dest = def;
+	}
+}
+
+static bool file_exists(const char *path) {
+	return path && access(path, R_OK) != -1;
+}
+
+static char *expand_path(char *path, bool realloc) {
+	if (path != NULL) {
+		wordexp_t p;
+		if (wordexp(path, &p, WRDE_UNDEF) == 0) {
```

Comment:
This doesn't work if a directory name contains spaces.

Reply:
What would be the solution to fix that? That's the way sway expands it's config path.

Reply:
done

Reply:
This replaces multiple spaces with a single one, and other whitespace characters with a space.

Reply:
This should solve the problem better. Libiniparser removes the quotes when parsing strings, so no path should be quoted, which does break wordexp. I first tried a solution using $IFS to disable separating on whitespaces as indicated [here](https://linux.die.net/man/3/wordexp). But that doesn't seemed to be implemented by the gnu version of libc. So I hope this behaves well on most implementations.

Reply:
Can I resolv this with the new implementation via getenv?

Reply:
sry wrong thread